### ISSUE TITLE
not compatible with TB 68.0

### ISSUE THREAD

**Issue by Pledkae** [RealRaven2000/FiltaQuilla#15]
Any chance for making the plugin compatible with ThunderBird 68.0 (plugin is disabled now)?
(TB v68.0; FiltaQuilla v1.5)

**Comment by RealRaven2000** [comment]
I had a look at that last weekend, and was shocked how much the changes in Thunderbird 68 affect FiltaQuilla. They removed a whole mechanism (xbl binding) that FiltaQuilla heavily depends on. Definitely want to rewrite it, but it's going to take a few weeks.

**Comment by RealRaven2000** [comment]
I have added branch (thunderbird68)[https://github.com/RealRaven2000/FiltaQuilla/tree/thunderbird68] for experimenting

**Comment by Kwaadpepper** [comment]
Thanks for the good work !

**Comment by peci1** [comment]
I've managed to convert one of my addons that was based off of FiltaQuilla code. Maybe you can get some inspiration there... I didn't figure out how to get rid of the XBL binding, but it seems it will still work for a few more versions.

https://github.com/peci1/mailing-list-filter/commit/aac098f5717a2401d497511ed39833ce57cc19b2

**Comment by RealRaven2000** [comment]
> I've managed to convert one of my addons that was based off of FiltaQuilla code. Maybe you can get some inspiration there... I didn't figure out how to get rid of the XBL binding, but it seems it will still work for a few more versions.
> 

very cool, I will definitely check it out! But did you manage to make the filter editor work with the new method? The whole point of using xbl bindings was to allow editing FIltaQuilla conditions / actions...

**Comment by peci1** [comment]
Yeah, it works in the filter editor. I was also surprised XBL still works, and they say it should be completely removed somewhere before 78. And I've also found a mention somewhere that with Custom elements, there's currently no easy way to replicate the behavior of -moz-bind in CSS... :( 

You're right, I could mention FiltaQuilla more explicitly, e.g. in the readme. I've retained all of the license blocks in the files I've copied and edited, so there are some mentions about FiltaQuilla and Kent James, but readme would be even more appropriate :)

**Comment by RealRaven2000** [comment]
> Yeah, it works in the filter editor. I was also surprised XBL still works, and they say it should be completely removed somewhere before 78. And I've also found a mention somewhere that with Custom elements, there's currently no easy way to replicate the behavior of -moz-bind in CSS... :(
> 
I had found some code on the Thunderbird dxr where they use 
 MozXULElement.parseXULToFragment to load the appropriate XML as text, but I am not sure if that is going to be the "final form" for next year. That is basically what I want to do with FiltaQuilla for now, if I can make it work.

Chances are they will have to rewrite the whole filter dialog as HTML5 content, and I bet they will stuff it into a browser tab. Personally I hate putting settings into a browser tab, modeless dialog windows are far superior for many applications that need context.

**Comment by peci1** [comment]
Well, parseXULToFragment is definitely one options, but I think it's not enough for the current implementation of the filter editor. If you wanted to avoid XBL, you'd need to monitor the parent menulist element's onchange and react in case custom element is selected... And I'm not sure if it's possible to hook an element's onchage method without actually throwing away the old callback...

**Comment by ptomulik** [comment]
Hi,
after few days of struggling I finally managed to make my "resubmit" extension to work under 68. It uses filterEditor in more or less similar way as filtaQuilla, so similar issues in my case. Regarding the css bindings (missing "ruleactiontarget-*" bindings), the only working way I found, was by a hack (monkey patching thunderbirds custom element). Example code below (it replaces corresponding "-moz-binding" bindings). The following file is included from FilterEditor's XUL overlay.

https://github.com/ptomulik/mozilla-resubmit/tree/2.0.0a1

```
/*global customElements */

{
    function getChildNode(type) {
        const elementMapping = {
            // These are my action IDs. Each line corresponds to one removed entry:
            // .ruleactiontarget[type="resubmit@ezamber.pl#actionFooBar"] {
            //      -moz-binding: url(...);
            //  }
            // in CSS file.
            "resubmit@ezamber.pl#actionSendNow": "ruleactiontarget-replyto",
            "resubmit@ezamber.pl#actionSendLater": "ruleactiontarget-replyto",
            "resubmit@ezamber.pl#actionCompose": "ruleactiontarget-replyto",
            "resubmit@ezamber.pl#actionSendNowMsg": "ruleactiontarget-replyto",
            "resubmit@ezamber.pl#actionSendLaterMsg": "ruleactiontarget-replyto",
            "resubmit@ezamber.pl#actionComposeMsg": "ruleactiontarget-replyto"
        };
        const elementName = elementMapping[type];
        return elementName ? document.createXULElement(elementName) : null;
    }

    function patchRuleactiontargetWrapper() {
        "use strict"

        let wrapper = customElements.get("ruleactiontarget-wrapper");
        if (wrapper) {
            let prevMethod = wrapper.prototype._getChildNode;
            if (prevMethod) {
                wrapper.prototype._getChildNode = function(type) {
                    let element = getChildNode(type);
                    return element ? element : prevMethod(type);
                };
            }
        }
    }

    patchRuleactiontargetWrapper();
}
```

Tested under 68.2.2.

**Comment by ptomulik** [comment]
In #20 I'm trying to resolve this issue.

**Comment by ptomulik** [comment]
Is anything left to be done for this? My guess is that the extension should work with the most recent changes. Release, maybe?

**Comment by RealRaven2000** [comment]
> Is anything left to be done for this? My guess is that the extension should work with the most recent changes. Release, maybe?

no, it turns out `document.getAnonymousNodes(this)` is failing at runtime, I am currently refactoring using standard DOM functions from the perspective of the XUL element, like this:

```
this.appendChild(MozXULElement.parseXULToFragment(`
  <hbox>
    <textbox class="ruleactionitem" onchange="this.parentNode.value = this.value;"></textbox>
    <toolbarbutton image="chrome://filtaquilla/skin/folder.png" class="focusbutton" tooltiptext="FROM-DTD-filebutton" oncommand="this.parentNode.parentNode.getURL()"></toolbarbutton>
  </hbox>
`));

this.hbox = this.getElementsByTagName("hbox")[0]; // document.getAnonymousNodes(this)[0];
this.textbox = this.hbox.firstChild;              //  document.getAnonymousNodes(this)[0].firstChild;
```

**Comment by RealRaven2000** [comment]
Also, it seems to not currently store the folder location (or the selection of "Save Attachments to") condition for me; investigating that. I don't fully understand how the filter dialog persists this stuff...

**Comment by RealRaven2000** [comment]
I think this is wrong:

`class FiltaQuillaRuleactiontargetBase extends MozXULElement { }`

It needs to be more than that.

**Comment by ptomulik** [comment]
> I think this is wrong:
> 
> `class FiltaQuillaRuleactiontargetBase extends MozXULElement { }`
> 
> It needs to be more than that.

Well, the "native" Thunderbird's ``ruleactiontarget-*`` bindnings were converted to custom elements exactly same way as you did. The ``ruleactiontarget-base`` binding disappeared and there is no base class for the new custom elements to replace it. I guess they somehow got rid of the base and all elements just inherit ``MozXULElement``.  I could do the same, but for flexibility decided to add (empty) intermediate class ``FiltaQuilla....Base``. Have you observed any bug symptoms?

**Comment by ptomulik** [comment]
Yup, guess everything is in place, and the "missing" code (base class), your'e looking for, is probably this: https://hg.mozilla.org/comm-central/rev/6e49a3ec00cd#l3.14

```diff
+const updateParentNode = (parentNode) => {
+  if (parentNode.hasAttribute("initialActionIndex")) {
+    let actionIndex = parentNode.getAttribute("initialActionIndex");
+    let filterAction = gFilter.getActionAt(actionIndex);
+    parentNode.initWithAction(filterAction);
+  }
+  parentNode.updateRemoveButton();
+};
```
this is active also for our custom ``ruleactiontarget-*`` elements.

**Comment by RealRaven2000** [comment]
I just committed my changes. where does this code go?

**Comment by RealRaven2000** [comment]
Also one more thing - are these lines just placeholders?  I guess I need to use getBundleString here...

      this.launchtitle = "FROM-DTD-filebutton";

**Comment by RealRaven2000** [comment]
ok, I added updateParentNode... it looks pretty good. Except the folder textbox oin "Save Attachments To" is empty when I load the editor, but the value is definitely persisted. Investigating ....

**Comment by ptomulik** [comment]
Most native "ruleactiontarget-*" elements seem to invoke this, bu (unfortunatelly), ``updateParentNode`` is a local function of thunderbird's ``searchWidgets.js`` (so unavailable to us - it's hidden in their local scope). The ``.closest(".ruleaction")`` method looks for a closest ancestor element of ``this`` having CSS class ``"ruleaction"``. Mostly the "internal" XUL elements that implement given rule action target are part of higher level XUL element that wraps the whole rule action (two elements - rule action type and rule action target).

**Comment by ptomulik** [comment]
Ouch, I'm wrong. Looks this function is actually available to others, so far so good. :)

**Comment by ptomulik** [comment]
> Also one more thing - are these lines just placeholders? I guess I need to use getBundleString here...
> 
> ```
>   this.launchtitle = "FROM-DTD-filebutton";
> ```

Well, no idea. Didn't touch it. :)

**Comment by ptomulik** [comment]
> > [..]
> 
> no, it turns out `document.getAnonymousNodes(this)` is failing at runtime, I am currently refactoring using standard DOM functions from the perspective of the XUL element, like this:
> 
>[...]
https://bugzilla.mozilla.org/show_bug.cgi?id=1596800

**Comment by RealRaven2000** [comment]
> 
> 
> > Also one more thing - are these lines just placeholders? I guess I need to use getBundleString here...
> > ```
> >   this.launchtitle = "FROM-DTD-filebutton";
> > ```
> 
> Well, no idea. Didn't touch it. :)

ok - I added localization, using the good ole' string bundle. As long as it's available to us.

I usually have a util.getBundleString( ) function for any international strings that I load at runtime, hopefully it's not going to be too hard when they deprecate string bundles. God knows what I am going to do when they remove XUL altogether - it's going to get ugly...

**Comment by RealRaven2000** [comment]
First Test Version - 2.0 prerelease 29 - see below. Make sure to unpack the zip file, then pull the xul file into addons in THunderbird 68. No warrantees. Please add prerelease specific bugs below.

[FiltaQuilla-2.0pre29.zip](https://github.com/RealRaven2000/FiltaQuilla/files/3941950/FiltaQuilla-2.0pre29.zip)

**Comment by kgraefe** [comment]
First of all, thanks for taking over that project. I once wrote a [JavaScript Action][2] to copy tags to follow-up e-mails.

The JavaScript editor does not work, though. I can select "JavaScript Action" in the filter settings and clicking the editor button opens a window to enter text, but that text is discarded when clicking "OK". The action's text stays at "undefined".

Debug Log:
```
FiltaQuilla 9:30:21.631  [52645 ms]   
fq_FilterEditor.js - start...
FiltaQuilla 9:30:21.631  [0 ms]   
fq_FilterEditor.js - Finished.
FiltaQuilla 9:30:21.653  [22 ms]   
Could not retrieve bundle string: filtaquilla.editJavascript
```

[2]: https://github.com/kgraefe/thunderbird-autotag

**Comment by RealRaven2000** [comment]
> First of all, thanks for taking over that project. I once wrote a [JavaScript Action](https://github.com/kgraefe/thunderbird-autotag) to copy tags to follow-up e-mails.
> 
> The JavaScript editor does not work, though. I can select "JavaScript Action" in the filter settings and clicking the editor button opens a window to enter text, but that text is discarded when clicking "OK". The action's text stays at "undefined".

Just tested, I think I fixed the bundle string problem, but there is a general problem of ptransporting back the script (or edited actions in general) into the textbox (same with save attachments / run program etc.). When I edit script directly in the JS Action textbox, I could actually make the filter work:

![image](https://user-images.githubusercontent.com/6314674/70527026-d09d9080-1b42-11ea-9387-e73b2ac8a30e.png)

Still need to figure out how the text control interacts with the filter storage mechanism, as it appears to store the info "behind the scenes" for the attachments commands.

**Comment by kgraefe** [comment]
Typing into the textbox has the same issue for me: it is "undefined" after closing the filter editor with "OK" and reopening it.

**Comment by RealRaven2000** [comment]
> 
> 
> Typing into the textbox has the same issue for me: it is "undefined" after closing the filter editor with "OK" and reopening it.

Yes - it's part of the funked up transfer of the value to the textbox field (it still works though). Anyway I found this, when I debugged into the `new updateParentNode()` function kindly provided by @Pledkae 

![image](https://user-images.githubusercontent.com/6314674/70531118-30e50000-1b4c-11ea-90e9-a46ded166d08.png)

so there is more code that copies from the value to the textbox, and that is called to early (before the call to `updateParentNode()` whereas it should be called after). New commit coming up shortly!

Current code:

![image](https://user-images.githubusercontent.com/6314674/70531286-85887b00-1b4c-11ea-8342-ed8172b27bd3.png)

**Comment by RealRaven2000** [comment]
So I think I solved some of the problems with the Javascript editor (including some layout stuff). haven't managed to fix the fact that the source code is center aligned.

Also, it acts kind of unexpected with newline characters - I can't edit any in and if I paste some formatted text from the clipboard it replaces the newlines with comma characters, here is what it looks like now:

![image](https://user-images.githubusercontent.com/6314674/70534120-48bf8280-1b52-11ea-9f66-d34881f80ec2.png)

Anyway, still feeling the progress, here is a trial version:
[FiltaQuilla-2.0pre35.zip](https://github.com/RealRaven2000/FiltaQuilla/files/3945089/FiltaQuilla-2.0pre35.zip)

**Comment by RealRaven2000** [comment]
As regards newlines, this comment from Kent James:

```
// The unicode line separator \u2028 is recognized by js as a line terminator,
//  but survives the storage in a filter editor file without getting
//  truncated. So we use it to store the newlines.
const LS = '\u2028';
```

probably doesn't hold true anymore :(

**Comment by kgraefe** [comment]
Newlines have have been a problem in the old version, too. I used to [remove them][1] before pasting the script into the filter. Otherwise it corrupted the filter settings file...

[1]: https://github.com/kgraefe/thunderbird-autotag/blob/master/Makefile

**Comment by kgraefe** [comment]
With FiltaQuilla-2.0pre35 editing works in the textbox as well as in the editor window. 👍 

But the filter does not run. (I assume `alert("hi");` should open some kind of message box saying hi?)

**Comment by ptomulik** [comment]
Copy-pasted the JS code (multiline one with alerts) directly into ruleationtarget input and tried. Seems to work (alert dialogs appeared). Pasting into textbox dialog yields a string with commas instead of newlines. FQ-2.0pre35.

**Comment by kgraefe** [comment]
Works for me as well, I had wrong filter conditions. 👍

**Comment by ptomulik** [comment]
> [..]. Pasting into textbox dialog yields a string with commas instead of newlines. FQ-2.0pre35.

This is still an error.

**Comment by RealRaven2000** [comment]
> Copy-pasted the JS code (multiline one with alerts) directly into ruleationtarget input and tried. Seems to work (alert dialogs appeared). Pasting into textbox dialog yields a string with commas instead of newlines. FQ-2.0pre35.

Yes but that's only half of the truth - the reason Kent replaced them because they can't be stored in msgFilterRules.dat. Remember they end up as a single line there:

`actionValue="alert(\"hoi there\"); for (let i=0; i<10; i++) { if (i>6) alert(i); }"`

**Comment by RealRaven2000** [comment]
Here is a nicer code example for Javascript testing:

`for (let i = 0; i < msgHdrs.length; i++) {  let hdr = msgHdrs.queryElementAt(i, Ci.nsIMsgDBHdr);  let txt = "subject is: " + hdr.subject;  alert(txt);}`

**Comment by ptomulik** [comment]
Looks like XUL texbox replaces newlines with commas at input by default. We can alter this behavior,  for example by changing this
```
<textbox id="jscode" flex="1" multiline="true" rows="10" />
```
to this
```
<textbox id="jscode" flex="1" multiline="true" newlines="pasteintact" rows="10" />
```
in jsEditor.xul.

**Comment by ptomulik** [comment]
There are also other possible values for the ``newlines`` attribute

https://developer.mozilla.org/en-US/docs/Archive/Mozilla/XUL/textbox

**Comment by RealRaven2000** [comment]
> There are also other possible values for the `newlines` attribute
> 
> https://developer.mozilla.org/en-US/docs/Archive/Mozilla/XUL/textbox
> https://developer.mozilla.org/en-US/docs/Archive/Mozilla/XUL/Attribute/newlines

Good find! I will probably go with newlines="pasteintact" for now. There is not much help in having ones code "destroyed" as soon as it hits that window. We can still sanitize at onAccept if needed.

**Comment by ptomulik** [comment]
Anyway, for some reason, the newlines are not preserved after closing the dialog.

**Comment by RealRaven2000** [comment]
> 
> 
> > There are also other possible values for the `newlines` attribute
> > https://developer.mozilla.org/en-US/docs/Archive/Mozilla/XUL/textbox
> > https://developer.mozilla.org/en-US/docs/Archive/Mozilla/XUL/Attribute/newlines
> 
> Good find! I will probably go with newlines="pasteintact" for now. There is not much help in having ones code "destroyed" as soon as it hits that window. We can still sanitize at onAccept if needed.

I just tried it and there is a ridiculous amount of whitespace between the lines. Still not quite sure what's happening here. maybe I will switch to <html:input type="text"> or something...

**Comment by ptomulik** [comment]
Well, I've seen this, but it doesn't look like whitespaces in the text, rather a display issues (still some textbox properties, maybe some CSS or such). If you try to walk with cursor, you'll probably realize that there is no intermediate lines/spaces between the lines containig text.

**Comment by RealRaven2000** [comment]
> Well, I've seen this, but it doesn't look like whitespaces in the text, rather a display issues (still some textbox properties, maybe some CSS or such). If you try to walk with cursor, you'll probably realize that there is no intermediate lines/spaces between the lines containig text.

Yeah I know, it's probably lineheight or related. Something in the multiline textbox behavior doing something wrong, maybe I am missing a thunderbird specific class.

**Comment by ptomulik** [comment]
No, look :)
https://developer.thunderbird.net/add-ons/tb68/changes#less-than-textbox-multiline-true-greater-than

**Comment by ptomulik** [comment]
In other words, you need this in jsEditor.xul
```
...
<dialog ...  xmlns:html="http://www.w3.org/1999/xhtml"   ...>
 ...
  <html:textarea id="jscode" flex="1"/>
</dialog>
```

Also, all "multiline" texbox in custom elements, if any, should be probably replaced with html:textarea.

**Comment by ptomulik** [comment]
I see mutline textboxes in ``options.xul``. They should be replaced with ``html:textarea``'s as well. Havent spot other occurrences.

**Comment by ptomulik** [comment]
Ok, #27 and #28 shall fix issues with textboxes in jsEditor and in options' dialog.

**Comment by RealRaven2000** [comment]
> In other words, you need this in jsEditor.xul
> 
> ```
> ...
> <dialog ...  xmlns:html="http://www.w3.org/1999/xhtml"   ...>
>  ...
>   <html:textarea id="jscode" flex="1"/>
> </dialog>
> ```
> 
> Also, all "multiline" texbox in custom elements, if any, should be probably replaced with html:textarea.

Yeah I used exactly that in my Add-on SmartTemplate4, html elements work quite well there. Not to sure whether there are any downsides specific to Filter actions.

**Comment by ptomulik** [comment]
> > In other words, you need this in jsEditor.xul
> > ```
> > ...
> > <dialog ...  xmlns:html="http://www.w3.org/1999/xhtml"   ...>
> >  ...
> >   <html:textarea id="jscode" flex="1"/>
> > </dialog>
> > ```
> > 
> > 
> > Also, all "multiline" texbox in custom elements, if any, should be probably replaced with html:textarea.
> 
> Yeah I used exactly that in my Add-on SmartTemplate4, html elements work quite well there. Not to sure whether there are any downsides specific to Filter actions.

Actually it's not a choice, but must. As [the 6168 migration guide](https://developer.thunderbird.net/add-ons/tb68/changes#less-than-textbox-multiline-true-greater-than) says, multiline textboxes have to be replaced with html textarea elements.

**Comment by RealRaven2000** [comment]
I have replaced the textboxes in options.xul with descriptions, looks decent now:

![image](https://user-images.githubusercontent.com/6314674/70569154-4b8d9800-1b91-11ea-9569-f52a3b7b6b1a.png)

> Actually it's not a choice, but a must. As the 61-68 migration guide says, multiline textboxes have to be replaced with html textarea elements.

I know :o) I was just dragging my heels with FiltaQuilla because I was dragging the xbl bindings replacement. If you check my other Add-ons (SmartTemplate⁴, QuickFolders, MenuOnTop) I have already replaced all multline textboxes (and color pickers, too). Eventually everything is going to be replaced with (X)HTML but work on the Thunderbird Core conversion itself is stil W.I.P.

**Comment by RealRaven2000** [comment]
Latest test version - 2.0 prerelease 51

[FiltaQuilla-2.0pre51.zip](https://github.com/RealRaven2000/FiltaQuilla/files/3947335/FiltaQuilla-2.0pre51.zip)

are we ready for uploading yet?
if you have any suggestions on the "Support" section, I would like to use some of the traffic for FiltaQuilla to grow my add-on quickFilters, and monetise through this instead:
![image](https://user-images.githubusercontent.com/6314674/70572947-c1493200-1b98-11ea-9df6-b4bd9346dfef.png)

Instead of linking to its donation page, I could alternatively link to the latest version history page, at:
http://quickfilters.mozdev.org/version.html#latest

**Comment by ptomulik** [comment]
I think it's worth being uploaded. Even if there are some regressions, it's still better for 68+ users to have something over nothing. The TB <=61 users shall not be altered as in their case 1.x versions get installed from add-on repo.

**Comment by RealRaven2000** [comment]
> I think it's worth being uploaded. Even if there are some regressions, it's still better for 68+ users to have something over nothing. The TB <=61 users shall not be altered as in their case 1.x versions get installed from add-on repo.

Okay, I will do some smoke tests today to make sure it mostly "does what it says on the tin". Would love if it could also drive some traffic to [[fuiqkfilters]](https://addons.thunderbird.net/en-US/thunderbird/addon/quickfilters/). Christmas sales are starting soon...

**Comment by RealRaven2000** [comment]
@Pledkae I really resent the fact that we are still using code within bindings.xml for conditions, xbl bindings are pretty much deprecated and it makes debugging much harder. Just don't like implemented code withing XML, can we not do the code for conditions in the JavaScript module as well? I am just debugging the "Tag" binding and find it quite cumbersome. :(

**Comment by RealRaven2000** [comment]
I have done some additional work on the tag conditions but I still cannot make the  value appear in the dropdown when the filter editor dialog opens, any pointers?

![image](https://user-images.githubusercontent.com/6314674/70638397-b8a43a80-1c30-11ea-9ef5-083edd22585d.png)

**Comment by RealRaven2000** [comment]
Well I tried removing the xul prefix in bindings.xml in order to access the missing selectedItem property (and seeing that the other dropdown are also plain `<menulist>s` and not `<xul:menulist>`

![image](https://user-images.githubusercontent.com/6314674/70644290-a5966800-1c3a-11ea-8d65-c38653c191ee.png)

well, that didn't work quite well either. Just wish I know more about this stuff >:(

**Comment by ptomulik** [comment]
> Well I tried removing the xul prefix in bindings.xml in order to access the missing selectedItem property (and seeing that the other dropdown are also plain `<menulist>s` and not `<xul:menulist>`
> 
> [...]
> well, that didn't work quite well either. Just wish I know more about this stuff >:(

Tag menus implemented in #32.

**Comment by RealRaven2000** [comment]
I managed another milestone, just now repaired the  "Tag ... " Search Terms after some very helpful refactoring by @ptomulik - the implementation (Javascript part) of the flags condition is now moved successfully into the filtaquilla.js module making coding and troubleshooting a lot easier. In the long term I want to eliminate most (if not all) code from the xml parts as this make code management and debugging a lot easier. Trial version below:

[FiltaQuilla-2.0pre68.zip](https://github.com/RealRaven2000/FiltaQuilla/files/3955675/FiltaQuilla-2.0pre68.zip)

(remember to unpack the archive before installing the xul file - github doesn't support attaching binaries)

**Comment by RealRaven2000** [comment]
I have juist released version 2.0, closing this issue.
v2. 0 Released on 13/12/2019

For new bugs / feature requests  pertaining to FiltaQuilla v2.0 please open separate issues going forward!
